src/services/createRoute.ts

```diff
@@ -3,48 +3,74 @@ import { CombineKey } from '@/services/combineKey'
 import { Path, ToPath, toPath } from '@/types/path'
 import { Query, ToQuery, toQuery } from '@/types/query'
 import { RouteMeta } from '@/types/register'
 import { Route } from '@/types/route'
+import { Identity } from '@/types/utilities'
 import { checkDuplicateKeys } from '@/utilities/checkDuplicateKeys'
 
+type ParentPath<TParent extends Route | undefined> = TParent extends Route ? TParent['path'] : Path<'', {}>
+type ParentQuery<TParent extends Route | undefined> = TParent extends Route ? TParent['query'] : Query<'', {}>
+
+type RouteParams<
+  TPath extends string | Path | undefined,
+  TQuery extends string | Query | undefined,
+  TParent extends Route | undefined = undefined
+> = ExtractParamTypes<Identity<CombinePath<ParentPath<TParent>, ToPath<TPath>>['params'] & CombineQuery<ParentQuery<TParent>, ToQuery<TQuery>>['params']>>
+
 export function createRoute<
   const TName extends string | undefined = undefined,
   const TPath extends string | Path | undefined = undefined,
   const TQuery extends string | Query | undefined = undefined,
-  const TMeta extends RouteMeta = RouteMeta,
-  TComponent extends Component | undefined = undefined
->(options: CreateRouteOptions<TName, TPath, TQuery, TMeta> & WithHooks & WithComponent<TComponent> & WithoutParent): Route<ToKey<TName>, Host<'', {}>, ToPath<TPath>, ToQuery<TQuery>, TMeta>
+  const TMeta extends RouteMeta = RouteMeta
+>(options: CreateRouteOptions<TName, TPath, TQuery, TMeta> & WithHooks & WithoutComponents & WithoutParent): Route<ToKey<TName>, Host<'', {}>, ToPath<TPath>, ToQuery<TQuery>>
 
 export function createRoute<
+  const TParent extends Route,
   const TName extends string | undefined = undefined,
   const TPath extends string | Path | undefined = undefined,
   const TQuery extends string | Query | undefined = undefined,
-  const TMeta extends RouteMeta = RouteMeta,
-  TComponents extends Record<string, Component> | undefined = undefined
->(options: CreateRouteOptions<TName, TPath, TQuery, TMeta> & WithHooks & WithComponents<TComponents> & WithoutParent): Route<ToKey<TName>, Host<'', {}>, ToPath<TPath>, ToQuery<TQuery>, TMeta>
+  const TMeta extends RouteMeta = RouteMeta
+>(options: CreateRouteOptions<TName, TPath, TQuery, TMeta> & WithHooks & WithoutComponents & WithParent<TParent>): Route<CombineKey<TParent['key'], ToKey<TName>>, Host<'', {}>, CombinePath<TParent['path'], ToPath<TPath>>, CombineQuery<TParent['query'], ToQuery<TQuery>>>
 
 export function createRoute<
+  TComponent extends Component,
+  const TName extends string | undefined = undefined,
+  const TPath extends string | Path | undefined = undefined,
+  const TQuery extends string | Query | undefined = undefined,
+  const TMeta extends RouteMeta = RouteMeta
+>(options: CreateRouteOptions<TName, TPath, TQuery, TMeta> & WithHooks & WithComponent<TComponent, RouteParams<TPath, TQuery>> & WithoutParent): Route<ToKey<TName>, Host<'', {}>, ToPath<TPath>, ToQuery<TQuery>>
```

Comment:
The rest of the changes to overloads are adding params to the component types. What use to be `WithComponent<TComponent>` is now `WithComponent<TComponent, RouteParams<TPath, TQuery>>`